File: __LEGACY__/ds-leetcode/longest-word-in-dictionary.java
class Solution {
    public String longestWord(String[] words) {
        // Trie data structure
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode current = root;
            for (char c : word.toCharArray()) {
                if (current.children[c - 'a'] == null) {
                    current.children[c - 'a'] = new TrieNode();
                }
                current = current.children[c - 'a'];
            }
            current.isWord = true;
        }

        // DFS
        String result = "";
        dfs(root, "", result);

        return result;
    }

    private void dfs(TrieNode root, String word, String result) {
        if (root.isWord) {
            if (word.length() > result.length() || (word.length() == result.length() && word.compareTo(result) < 0)) {
                result = word;
            }            
        }
        for (int i = 0; i < 26; i++) {
            if (root.children[i] != null) {
                dfs(root.children[i], word + (char)(i + 'a'), result);
            }
        }
    }

}

File: __LEGACY__/ds-leetcode/longest-word-in-dictionary.py
class Solution:
    def longestWord(self, words: List[str]) -> str:
        words.sort()
        longest = ''
        for word in words:
            if len(word) > len(longest) or (len(word) == len(longest) and word < longest):
                if self.is_valid(word, words):
                    longest = word
        return longest

    def is_valid(self, word, words):
        if word == '':
            return True
        for w in words:
            if w.startswith(word):
                return self.is_valid(w, words)
        return False